### ISSUE TITLE
Question: Postgres Database client_metrics table size

### ISSUE THREAD

**Issue by Green-Apple** [Unleash/unleash#890]
We recently ran into size issues with our Postgres database and noticed the `client_metrics` table was incredibly large

```
Top 10 table Using space
 table_schema |     table_name      | total_size | data_size  | external_size
--------------+---------------------+------------+------------+---------------
 public       | client_metrics      | 10141 MB   | 9235 MB    | 906 MB
 public       | events              | 5336 kB    | 5064 kB    | 272 kB
 public       | client_instances    | 400 kB     | 288 kB     | 112 kB
 public       | features            | 288 kB     | 200 kB     | 88 kB
 public       | client_applications | 88 kB      | 32 kB      | 56 kB
 public       | context_fields      | 64 kB      | 8192 bytes | 56 kB
 public       | strategies          | 64 kB      | 8192 bytes | 56 kB
 public       | settings            | 64 kB      | 8192 bytes | 56 kB
 public       | migrations          | 56 kB      | 8192 bytes | 48 kB
 public       | tag_types           | 32 kB      | 8192 bytes | 24 kB
```

at 10GB in size and growth continuing to accelerate as server usage continues to grow. Nearly 1GB of metrics a month at current rate and climbing. While we can manually go into the DB and purge old rows, is there any general support for cleanup of metrics without us creating our own chron job to clean the table? Certainly from an Unleash Server UI perspective, metrics aren't visually shown for data older then an hour. 

We are continuing to do analysis on our end to determine what exactly is going on with our server and the metrics table, but want to throw this out early in case anyone else had experienced similar issues.

**Comment by Green-Apple** [comment]
So from examining the database, we are only seeing records for the past hour. Does this mean the Unleash server is only keeping rows in the table for an hour. If so, that means given our scale and the way metrics are being stored, we are generating >10GB an hour in metric data. Seems like a lot when the rest of the database combined is <10MB.

**Comment by ivarconr** [comment]
Yes this sounds insane. To me it sounds like you have configured something wrong at the SDKs. All SDKs will register when the client application starts, and send aggregated usage metrics every 30s. These are stored for an hour. 

I have sometimes seen wrong usages where the SDK ends up being instansiated per request. This will lead to an extreme load of metrics. 

Unleash do expose a Prometheus endpoint so you can try to leverage that to see number of request and registrations.

**Comment by Green-Apple** [comment]
So we did analysis on the major applications and the metric table. None of them were creating new SDK clients on requests, but at least one legacy spring application seems to be registering more clients the we'd expect though it seems to be non-growing amount per application, i.e. like 20 per application, but not increasing. 

| Application                    | Rows  | Estimated Active Servers Connected to Unleash |
|--------------------------------|-------|-----------------------------------------------|
| Legacy Monolithic App       | 37135 | ~620 (Doesn't seem realistic)                    |
| Monolithic Server                   | 17793 | ~297                                           |
| All other microservices  | 10,235   | ~170                                          |

So the legacy mono app, i'm not sure how many instances are deployed, but I can't imagine it being more then 100, so 620 definitely represents an issue on our end. That server only has like two feature flags, so the JSON are pretty small.

The microservices also have relatively small JSON given most of them have < 5 flags.

The other mono server though has a large number of active feature flags, around 70 currently. Which is making the JSON relatively large ~2k characters. Also, the ~300 servers is a relatively reasonable number, higher than I would expect, but very possible given our infrastructure.

So I guess my question is, beyond just scaling the unleash database up (which we've already done moving to 20GB with dynamic scaling allowed up to 1TB) is their any other optimizations or improvements that we could be doing. Additionally, are there better ways to store metric information that involves less storage or less data being sent? 

I don't anticipate our unleash server demand decreasing, if anything, we are looking at 10-50x fold increase in feature flag usage as our multi-clusters start coming online. We already are spooling up additional unleash servers, but the sheer volume of data >10GB of metric data an hour and growing seems high.

**Comment by ivarconr** [comment]
Thanks for sharing this, its really valuable! I have used Unleash in large environments, and also know about a few where they have  hundreds of feature toggles and thousands of connected application instances. I have never seen these databases above a few GB, so I am still a bit puzzled to be honest. I still think there is something odd going on and would love the opportunity to explore some of the data, if that's possible. 

> So I guess my question is, beyond just scaling the unleash database up (which we've already done moving to 20GB with dynamic scaling allowed up to 1TB) is their any other optimizations or improvements that we could be doing. 
> Additionally, are there better ways to store metric information that involves less storage or less data being sent?

You could increase the metrics interval on the client SDK side. It is also possible to disable the metrics all together (at least in test environments this can make sense if there is much auto-generated data)

Also Unleash has computes and keep the actual counts in memory and only uses the client_metrics table for replication across instances and read that table in case of a reboot. We could in theory keep the raw metrics only for synchronization and instead store the calculated views for last minute and last hour. 

Metrics is definitely an improvement point in Unleash.

**Comment by Green-Apple** [comment]
Thanks Ivar, what data would you be interested in seeing. I would need to scrub some of it, but I could probably get you whatever information you'd like. I think for test, we are going to try reducing the metric frequency, but disabling it could result in product teams having a harder time verifying feature metrics.

**Comment by ivarconr** [comment]
Yes, pleas take away any internal data as needed. I think a few examples of registrations for the various apps and a couple of actual client_metrics rows across the the apps would be beneficial. I assume they are pretty similar per client application. I seek enough data to be able to simulate your setup and see whats going on. 

Also; could verify that delete actually works (no client_metrics entries older than approx 1 hour).

**Comment by Green-Apple** [comment]
I will work on getting that data Ivar, but it will probably be next week before I can upload the information. We did confirm that no entries are older than hour, so at least we have that going for us.

**Comment by ivarconr** [comment]
This is really valuable. Feel free to share it with me directly [on slack](https://join.slack.com/t/unleash-community/shared_invite/enQtNjUxMjU2MDc0MTAxLTJjYmViYjkwYmE0ODVlNmY1YjcwZGRmZWU5MTU1YTQ1Nzg5ZWQ2YzBlY2U1MjlmZDg5ZDRmZTMzNmQ5YmEyOGE) if you prepare not to share it to everyone on internet.

**Comment by Green-Apple** [comment]
This ended up being a little on back burner, but here are some metrics

In context, this is for our TEST environment unleash server. We have a PROD unleash server as well that has considerably less size issues though certainly not ~1 GB in database size. 

| Application     | Rows  | Sum(Size of Row) |
|-----------------|-------|------------------|
| Microservice A  | 670   | 105860           |
| Legacy Mono App | 42583 | 7325833          |
| Microservice B  | 609   | 113881           |
| "New" Mono App  | 16488 | 2959729          |

On thing that became apparent to me is that something isn't configured correctly for the Legacy Mono App. It's not generating a new client per request, but each JavaEE Spring service/EJB module context is spinning up its own copy instead of a shared client for the entire deployable. There are only two feature flags in the codebase and they control batch jobs, so every single one of those 42,583 rows is basically an empty JSON. So we can immediately clear out ~70% of the size issue just be fixing that.

I also confirmed no rows older then ~1h so that isn't an issue.

For the "New" Mono App, we had ~300 instances registered though I discovered that our CI/CD pipelines were spinning up client instances as well while running the test suite. So some fixes I ended up doing including disable metric send when client is initialized in pipeline and reduced metric send frequency to every two minutes instead of every one minute for test environments. Production is still on 1 min interval. I anticipate this to reduce overall size by 55-60% which given these two solutions should get our total test environment DB size to closer to 1-2GB which while still relatively large, is much more manageable.

"New" Mono App registration logs from client
```
    [2021-07-20 15:43:53,251 CDT] LOG_LEVEL="INFO" ENVIRONMENT="testing" SERVER="xblmbat1p12.abc.abcdefghij.abc" CATEGORY="FeatureToggleLogger" THREAD_ID="unleash-api-executor" TRACEABILITY_ID="" *********** UnleashDefaultSubscription : Client has been registered. : AppName=LEGACY MONO APP : InstanceId=testing.xblmbat1p12.abc.abcdefghij.abc : Interval=60 : SdkVersion=unleash-client-java:3.3.3 : Started=2021-07-20T20:43:52.807 : Strategies=[gradualRolloutSessionId, flexibleRollout, IncrementalRolloutStrategy, *******NumberRolloutStrategy, applicationHostname, gradualRolloutUserId, ReleaseDateRolloutStrategy, gradualRolloutRandom, *******RolloutStrategy, *********RolloutStrategy, default, userWithId, *********RolloutStrategy, remoteAddress, EnvironmentRolloutStrategy]
```

In terms of the rows themselves, the largest JSONs have ~25 active feature flag entries and have size between 1800-2000 character, but a large number of these rows are also pretty empty. (Given that this is TEST scoped, not surprising).

In general though, I do wonder if there is a more effective (space wise) mechanism to store these metrics and improve metrics in general. My company also uses another software as feature flagging tool that I've found superior for Business Analytics and metric gathering. I have found Unleash to be superior tool from a technical implementation perspective and our product teams use it often for technical rollouts. However, for customer research perspective, I've personally found the metric gathering and storage of the other tool far more helpful in making business decisions regarding what our customers ultimately need. I would be happy to go into some suggestions into what would make this tool more attractive from a business perspective if that is of interest to you.

**Comment by ivarconr** [comment]
(Sorry, I have been a few weeks off for some holiday this a bit late reply from my end)

Thanks for talking the time and effort. 

>  but each JavaEE Spring service/EJB module context is spinning up its own copy instead of a shared client for the entire deployable. 

I think there could be things we could look in to in our SDK in order to detect if the SDK is instantiated multiple times. At least we could warn when this happens. (E.g. For the Node.js SDK we also have an initialize function which will grantee to only create a single instance). 

> I would be happy to go into some suggestions into what would make this tool more attractive from a business perspective if that is of interest to you.

That would be nice. Today metrics is primarily used to debug the usage of Unleash. We want to expand this use case in coming releases (e.g. per app, environment, context fields etc). At some point we also want Unleash to support other type of metrics (technical quality, aka SLO, but also business metrics). 

Could you connect with me on ivar[at]getunleash.io and suggest a few timeslots that could work for you next week?

**Comment by Green-Apple** [comment]
So unrelated to the previous comment, we had our unleash server go down again due to DB size issues, it had exceeded 20GB in size. To help free up space, we attempted to delete the `client_metrics` table, but still the size was showing as >20 GB despite `select * from client_metrics` queries to the table showing no rows. We ended up doing a [Postgresql Vacuum](https://www.postgresql.org/docs/9.1/sql-vacuum.html) command `VACUUM(FULL, ANALYZE, VERBOSE) client_metrics;` which took quite a while to run, but once it finished, the table showed around 12-15MB in size, a significant improvement in size and has been relatively constant in size this yesterday when the vacuum command was run

As to why the table was not properly purging and releasing the space of deleted rows, we haven't been table to dig into yet. But for now, it seems to be resolved.

**Comment by ivarconr** [comment]
We will address this in the #908 where we will store the metrics quite differently. We will basically aggregate metrics in hour buckets on a few selected metrics.

**Comment by ivarconr** [comment]
Closing this as we are working on a new way of handling metrics in #908. The new way will reduce metrics to hour buckets per application and environment for all metrics.
